collector/src/main.rs

```diff
@@ -1141,7 +1164,99 @@ fn main_result() -> anyhow::Result<i32> {
 
             Ok(0)
         }
+        Commands::Download(opts) => {
+            let crate_path = match opts.command {
+                DownloadSubcommand::Git { url } => {
+                    download_from_git(&benchmark_dir, opts.name.as_deref(), &url)?
+                }
+                DownloadSubcommand::Crate { krate, version } => {
+                    download_from_crates_io(&benchmark_dir, &krate, &version, opts.name.as_deref())?
+                }
+            };
+            prepare_downloaded_benchmark(&crate_path);
+            Ok(0)
+        }
+    }
+}
+
+fn download_from_git(
+    benchmark_dir: &Path,
+    name: Option<&str>,
+    url: &str,
+) -> anyhow::Result<PathBuf> {
+    let target = PathBuf::from(benchmark_dir);
+    let crate_name = match name {
+        Some(name) => name,
+        None => url
+            .trim_end_matches("/")
+            .trim_end_matches(".git")
+            .split("/")
+            .last()
+            .expect("Crate name could not be determined from git URL"),
+    };
+    let target = target.join(crate_name);
+    if target.exists() {
+        return Err(anyhow::anyhow!(
+            "Directory {} already exists",
+            target.display()
+        ));
+    }
+
+    Command::new("git")
+        .arg("clone")
+        .arg(url)
+        .arg(&target)
+        .output()
+        .expect("Git clone failed");
+    Ok(target)
+}
+
+fn download_from_crates_io(
+    benchmark_dir: &Path,
+    krate: &str,
+    version: &str,
+    name: Option<&str>,
+) -> anyhow::Result<PathBuf> {
+    let full_name = format!("{krate}-{version}");
+    let name = name.unwrap_or(&full_name);
+    let target = PathBuf::from(benchmark_dir).join(name);
+    if target.exists() {
+        return Err(anyhow::anyhow!(
+            "Directory {} already exists",
+            target.display()
+        ));
+    }
+
+    let url = format!("https://crates.io/api/v1/crates/{krate}/{version}/download");
+    let body = reqwest::blocking::get(url).expect("Cannot download crate");
+    let data = flate2::read::GzDecoder::new(body);
+    let mut archive = tar::Archive::new(data);
+
+    let tmpdir = tempfile::TempDir::new().unwrap();
+    archive.unpack(&tmpdir)?;
+    let unpacked_dir = std::fs::read_dir(&tmpdir.path())
+        .unwrap()
+        .next()
+        .unwrap()
+        .unwrap()
+        .path();
+    execute::rename(&unpacked_dir, &target)?;
+
+    Ok(target)
+}
+
+fn prepare_downloaded_benchmark(path: &Path) {
+    let manifest_path = path.join("Cargo.toml");
+    let manifest = std::fs::read_to_string(&manifest_path).expect("Cannot open Cargo.toml");
+    if !manifest.contains("[workspace]") {
+        let mut file = File::options().append(true).open(&manifest_path).unwrap();
+        writeln!(file, "[workspace]").expect("Cannot append workspace to Cargo.toml");
```

Comment:
I'm not sure whether unconditionally enabling workspaces like this is completely harmless, or would somehow change cargo behaviour when compiling ? In any case, you can also `cargo generate-lockfile` in your temp dir above before moving it to the collector benchmarks.

Reply:
That's actually a pretty good idea :) I'll do it like that, thanks.

Reply:
> That's actually a pretty good idea :) 

I'm used to manually adding benchmarks ^^

Reply:
can you also check what regenerating the lockfile does when a lockfile already exists ? or if we need to add e.g. the --offline or --frozen or other parameters: especially if there was another semver-compatible version released since then. If it does nothing great, if not maybe we should only regenerate it if it's not already present rather than unconditionally.

Reply:
Generating the lockfile twice seems to be ok, it doesn't produce an error and the lockfile stays there. I'm not sure what do you mean by "released since then". Currently the command assumes that it downloads a new crate that does not exist locally yet. It does not support overwriting or updating an existing benchmark (although I can add support for that if you like).

Reply:
If by "stays there", you mean that it is a no-op and the lockfile isn't changed, then we're good. We should probably use the --frozen parameter and so on to ensure this, mark will know more about this than I do. I'd just want to make sure we're not inadvertently benchmarking a slightly different version of the code (but that probably can't be avoided when the lockfiles are missing anyways) just as a precaution.

Reply:
Aah, now I understand what did you mean. So generating the lockfile actually changes it if it existed before which is problematic. But using `--frozen` doesn't help, because it then cannot update the registry. I think that the easiest thing would be just to skip generating the lockfile if it already exists, as you have suggested.

Reply:
I think `cargo metadata` or some other read-only but otherwise no-op command is the best way to just ensure the lockfile exists but otherwise not do anything.

I'm not aware of a way today to get cargo generate-lockfile to avoid bumping dependency versions.

That said, it's worth noting that crates.io packages, even if they have a lockfile, typically aren't making *use* of that lockfile -- only the root workspace lockfile matters for Cargo package resolution. So we could probably just use generate-lockfile. The main downside there is that if there *is* a lockfile, then we're making builds less reproducible, but since the lockfile isn't something we can typically depend on anyway I'm not sure it matters.

Reply:
`cargo metadata` sounds good. I would rather preserve the existing lock files of binary crates without updating them.

Reply:
I've believe I've seen `cargo metadata` update lockfiles before, but it probably doesn't matter indeed.

Reply:
Yeah, if it has to, it will (you could do --locked, I guess, but doesn't seem worth it). But it should by default leave the dependencies in the same versions -- it might update the lockfile to a newer format, though.